elastic/kibana
# feat(slo): Handle updating SLO

## 📝 Summary

Related to https://github.com/elastic/kibana/issues/141283

This PR introduces the slo revision and uses the tuple (slo.id, slo.revision) when creating the related Transform and aggregating the data.
It also introduces the route to update an existing slo along the application service handling the update.

As explained in the related [story](https://github.com/elastic/kibana/issues/141283), an edit can be breaking or non breaking. 
For non-breaking changes, i.e. editing anything but the indicator: we only need to update the saved object without bumping the slo revision.
For breaking changes, i.e. changing the indicator: we need to remove the previously aggregated data and transform (could become optional later with an extra parameter if we want to retain the previous data), create a new transform and start it.


## ✅ Manual testing


1. Create a new slo
2. Check revision in kibana saved object, expect revision === 1
3. Update the slo without breaking changes (change the name, time_window, objective...)
4. Get the slo 
5. Assert
   1. SLO data has been updated
   2. Transform is still running
   3. Check revision in kibana saved object, expect revision === 1
3. Update the slo with breaking changes (change the indicator's environment for example)
4. Get the slo 
5. Assert
   1. SLO data has been updated, including indicator
   2. Previous transform (named `slo-{id}-1`) has been deleted
   3. New transform (named `slo-{id}-2`) is running
   4. Check revision in kibana saved object, expect revision === 2
 


### ⚙️  Useful requests

<details><summary>Create SLO</summary>

```
curl --request POST \
  --url http://localhost:5601/mgx/api/observability/slos \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"name": "My SLO Availability",
	"description": "99% o11y-app all services availablility",
	"indicator": {
		"type": "slo.apm.transaction_error_rate",
		"params": {
			"environment": "development",
			"service": "o11y-app",
			"transaction_type": "request",
			"transaction_name": "GET /flaky",
			"good_status_codes": ["2xx", "3xx"]
		}
	},
	"time_window": {
		"duration": "7d",
		"is_rolling": true
	},
	"budgeting_method": "occurrences",
	"objective": {
		"target": 0.95
	}
}'
```
</details>


<details><summary>Update SLO (without breaking changes)</summary>

```
curl --request PUT \
  --url http://localhost:5601/mgx/api/observability/slos/7a42fe70-4022-11ed-93fe-659ac997b883 \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"name": "My Updated SLO",
	"time_window": {
		"duration": "30d",
		"is_rolling": true
	}
}'
```
</details>


<details><summary>Update SLO (with breaking changes)</summary>

```
curl --request PUT \
  --url http://localhost:5601/mgx/api/observability/slos/7a42fe70-4022-11ed-93fe-659ac997b883 \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"name": "Updated SLO bis",
	"indicator": {
		"type": "slo.apm.transaction_error_rate",
		"params": {
			"environment": "production",
			"service": "o11y-app",
			"transaction_type": "*",
			"transaction_name": "GET /flaky",
			"good_status_codes": [
				"2xx",
				"3xx",
				"4xx"
			]
		}
	}
}'
```
</details>



<details><summary>Get SLO</summary>

```
curl --request GET \
  --url http://localhost:5601/mgx/api/observability/slos/7a42fe70-4022-11ed-93fe-659ac997b883 \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'kbn-xsrf: oui'
```
</details>
